test/metabase/pulse/render/common_test.clj

```diff
@@ -29,29 +31,48 @@
+        (is (= "Af12,346" (fmt {::mb.viz/currency "AFN"})))
+        (is (= "₡12,346" (fmt {::mb.viz/currency "CRC"})))
+        (is (= "ZK12,346" (fmt {::mb.viz/currency "ZMK"})))
         (is (= "12,345.54 Cape Verdean escudos"
-               (fmt {::mb.viz/currency "CVE"
+               (fmt {::mb.viz/currency       "CVE"
                      ::mb.viz/currency-style "name"}))))
       (testing "Understands name, code, and symbol"
         (doseq [[style expected] [["name" "12,345.54 Czech Republic korunas"]
                                   ["symbol" "Kč12,345.54"]
                                   ["code" "CZK 12,345.54"]]]
-          (is (= expected (fmt {::mb.viz/currency "CZK"
+          (is (= expected (fmt {::mb.viz/currency       "CZK"
                                 ::mb.viz/currency-style style}))
               style))))
     (testing "scientific notation"
       (is (= "1.23E4" (fmt {::mb.viz/number-style "scientific"})))
-      (is (= "1.2346E4" (fmt {::mb.viz/number-style "scientific"
-                              ::mb.viz/decimals 4})))
+      (is (= "1.23E4" (fmt {::mb.viz/number-style "scientific"
+                            ::mb.viz/decimals     2})))
+      (is (= "1.235E4" (fmt {::mb.viz/number-style "scientific"
+                             ::mb.viz/decimals     3})))
+      (is (= "1.23455E4" (fmt {::mb.viz/number-style "scientific"
+                               ::mb.viz/decimals     5})))
       (is (= "1E4" (fmt {::mb.viz/number-style "scientific"
-                         ::mb.viz/decimals 0}))))
+                         ::mb.viz/decimals     0}))))
     (testing "Percentage"
       (is (= "1,234,554.32%" (fmt {::mb.viz/number-style "percent"})))
       (is (= "1.234.554,3200%"
-             (fmt {::mb.viz/number-style "percent"
-                   ::mb.viz/decimals 4
-                   ::mb.viz/number-separators ",."}))))
+             (fmt {::mb.viz/number-style      "percent"
+                   ::mb.viz/decimals          4
+                   ::mb.viz/number-separators ",."})))
+      (is (= "10%" (format 0.1 {::mb.viz/number-style "percent"})))
+      (is (= "1%" (format 0.01 {::mb.viz/number-style "percent"})))
+      (is (= "0.00001%" (format 0.0000001 {::mb.viz/number-style "percent"}))))
+    (testing "Match UI behavior for decimal values with no column formatting present"
+      (is (= ["2"    "0"]      [(format 2 nil) (format 0 nil)]))
+      (is (= ["2.1"  "0.1"]    [(format 2.1 nil) (format 0.1 nil)]))
+      (is (= ["2.01" "0.01"]   [(format 2.01 nil) (format 0.01 nil)]))
+      (is (= ["2"    "0.001"]  [(format 2.001 nil) (format 0.001 nil)]))
+      ;; Notice that (BigDecimal. 2.005) -> 2.0049999999... etc. so we have precision problems unless we
+      ;; send a BigDecimal in right away, which, as far as I can tell, is what we'll get from the query processor
+      (is (= ["2.01" "0.005"]  [(format 2.005M nil) (format 0.005 nil)]))
```

Comment:
The bigdecimal i think is a red herring here. Just pick a number that has an exact representation in binary rather than using BigDecimal as a way to hold the exact. Test bigdecimal as you want to test it but don't use it as a crutch around floating point weirdness. just get a better number.